### PR TITLE
UCM2: Intel: sof-hda-dsp: Control SOF processing from UCM

### PR DESCRIPTION
**Pull request by singalsu**
This is a set of patches to control with UCMv2 audio processing in Sound Open Firmware (SOF). The settings are generic, mostly pass-through mode, except for speaker where a generic high-pass filter and DRC boost are applied.

All the filters can be set up with a bespoke tuned version per product for a better audio experience.

### CODE REVIEW

**Review by perexg — CHANGES_REQUESTED** [review]

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+If.SOFIPCVer {
+	Condition {
+		Type ControlExists
+		Control "name='Post Mixer Analog Playback IIR Eq bytes'"
+	}
+	True.Define {
+	      SOFIPCVer "ipc4"
+	      BlobPath "${ConfTopDir}/sof/${var:SOFIPCVer}"
+	      PostMixerAnalogPlaybackIIRBytes "Post Mixer Analog Playback IIR Eq bytes"
```
[the remark this answers is not in the document]
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:13]: Thanks a lot for all the helpful suggestions! I hope to be able to share a new version next week some time.

[the remark this answers is not in the document]
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:13]: I'm trying this but I get "ALSA lib ucm_cond.c:378:(if_eval_one) condition block expected (If)". What could be wrong?

```
If.SOFIPCVer {
	Condition {
		Type ControlExists
		Control "name='Post Mixer Analog Playback IIR Eq bytes'"
	}
	True.Define {
	      SOFIPCVer "ipc4"
	      PostMixerAnalogPlaybackIIRBytes "Post Mixer Analog Playback IIR Eq bytes"
	      PostMixerAnalogPlaybackFIRBytes "Post Mixer Analog Playback FIR Eq bytes"
	      PostMixerAnalogPlaybackDRCBytes "Post Mixer Analog Playback DRC bytes"
	      PostMixerAnalogPlaybackDRCSwitch "Post Mixer Analog Playback DRC switch"
	}
	False.Define {
	      SOFIPCVer "ipc3"
	      PostMixerAnalogPlaybackIIRBytes "EQIIR1.0 eqiir_coef_1"
	      PostMixerAnalogPlaybackFIRBytes "EQFIR1.0 eqfir_coef_1"
	      PostMixerAnalogPlaybackDRCBytes "not available"
	      PostMixerAnalogPlaybackDRCSwitch "not available"
	}
}

If.SOFPath {
	Condition { Type AlwaysTrue }
	True.Define {
		BlobPath "${ConfTopDir}/blobs/sof/${var:SOFIPCVer}"
		SpeakerIirBlob "${var:BlobPath}/eq_iir/highpass_100hz_0db_48khz.blob"
		SpeakerFirBlob "${var:BlobPath}/eq_fir/pass.blob"
		SpeakerDrcBlob "${var:BlobPath}/drc/speaker_default.blob"
		HeadphoneIirBlob "${var:BlobPath}/eq_iir/pass.blob"
		HeadphoneFirBlob "${var:BlobPath}/eq_fir/pass.blob"
		HeadphoneDrcBlob "${var:BlobPath}/drc/passthrough.blob"
	}
}
```

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                False.If.endpoint_with_eq {
+                        Condition {
+                                Type ControlExists
+                                Control "name='${var:PostMixerAnalogPlaybackFIRBytes}'"
+                        }
+                        True {
+                                EnableSequence [
+                                        cset-tlv "name='${var:PostMixerAnalogPlaybackIIRBytes}' ${var:EndpointIirBlob}"
+                                        cset-tlv "name='${var:PostMixerAnalogPlaybackFIRBytes}' ${var:EndpointFirBlob}"
+                                        #shell "/bin/echo '${var:__endpoint} ${var:EndpointIirBlob} ${var:EndpointFirBlob}' >> /tmp/alsa-ucm.txt"
```
Comment (kv2019i) [inline ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:101]: I know it's not common to leave such commented-out code, but I did find these immensely useful when debugging and I could imagine these are useful for anyone who wants to add tuning for new laptop configurations later one. So if possible, +1 to leave these as examples.

**Review by kv2019i — COMMENTED** [review]
I tested an earlier version, but did another test with this version on a Intel IPC4 laptop. I tested specifying tuning blobs, switching between speaker and headphones, and also specifying the blob via a user config. All worked as advertise.

I can do a similar test when @singalsu you update the version.

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:13]: @perexg Please ignore above question, it's correct. The problem is another fix later in not duplicating the entire speaker and headphone Ifs.

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+			EnableSequence [
+				cset "name='Headphone Playback Switch' on"
+			]
+			DisableSequence [
+				cset "name='Headphone Playback Switch' off"
+			]
+		}
+	}
+
+        Macro.headphone.SofControl "endpoint=Headphone drcswitch=off"
```
[the remark this answers is not in the document]
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:127]: This change works, 

```
SectionDevice."Headphones" {
	Macro.headphone.SofControl "endpoint=Headphone drcswitch=off"
}
```

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+					cset "name='Speaker Playback Switch' on"
+				]
+
+				DisableSequence [
+					cset "name='Speaker Playback Switch' off"
+				]
+			}
+		}
+
+                Macro.speaker.SofControl "endpoint=Speaker drcswitch=on"
```
[the remark this answers is not in the document]
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:178]: But here, both of these fail. @perexg What is the correct way?

```
SectionDevice."Speaker" {
        Macro.speaker.SofControl "endpoint=Speaker drcswitch=on"
}

```

```
If.spk {
	Condition {
		Type ControlExists
		Control "name='Speaker Playback Switch'"
	}
	True.SectionDevice."Speaker" {
               Macro.speaker.SofControl "endpoint=Speaker drcswitch=on"
        }
}
```

The speaker controls are not set, there's no print Speaker happening to /tmp/alsa-ucm.txt and I see in /var/log/syslog message " wireplumber[3991]: Failed to enable ucm device Speaker".

[the remark this answers is not in the document]
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:178]: @pereg Sorry for noise, again my bad. In my testing version I had device specific customization in place but needed EQ blobs were not in place. I should get this now working!

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf (hunk 2 above)
Comment (singalsu) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf]: I moved the commented shell up before controls set to see the print before fails with them. There's now also a comment about use for debugging.

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+					#shell "/bin/echo '${var:__endpoint} ${var:EndpointIirBlob} ${var:EndpointFirBlob}' >> /tmp/alsa-ucm.txt"
+					cset-tlv "name='${var:PostMixerAnalogPlaybackIIRBytes}' ${var:EndpointIirBlob}"
+					cset-tlv "name='${var:PostMixerAnalogPlaybackFIRBytes}' ${var:EndpointFirBlob}"
+				]
+			}
+		}
+	}
+}
+
+# Merge this add to Headpones subtree in HDA/HiFi-analog.conf
```
Comment (ujfalusi) [inline ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:109]: Drop the 'add' from the comment

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+SectionDevice."Headphones" {
+	Macro.headphone.SofControl "endpoint=Headphone drcswitch=off"
+}
+
+# Merge this to Speaker subtree in HDA/HiFi-analog.conf
+SectionDevice."Speaker" {
+	Macro.speaker.SofControl "endpoint=Speaker drcswitch=on"
+}
```
Comment (ujfalusi) [inline ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:117]: Will this create a "Speaker" Section Device when there is no Speaker, like on upx-i11?

**Review by ujfalusi — COMMENTED** [review]
@singalsu, this looks clean to my untrained eyes, I only have one nitpick and one question

File: ucm2/Intel/sof-hda-dsp/HiFi-sof.conf (hunk 6 above)
Comment (perexg) [reply ucm2/Intel/sof-hda-dsp/HiFi-sof.conf:117]: Yes, the device will be always defined in this case. Move those lines to `If` block to use it only when the `Speaker Playback Switch` control exists.

**Review by perexg — CHANGES_REQUESTED** [review]
Except the speaker block (missing condition) all looks really nice now.

**Review by ujfalusi — APPROVED** [review]
@singalsu, looks clean to me.

**Comment by perexg** [comment]
Thanks. Applied now.

**Pull request closed without merging**